### DISCUSSION TITLE
ug --index doesn't use index - why?

### DISCUSSION THREAD

**Discussion by marcin-github** [Genivia/ugrep#429] [Q&A] [answered]
Hi,
I'm playing with ug-indexer and `ug --index`.  I have 11GB text file, I indexed it:
```
ugrep-indexer -v -9

> index accuracy: 9 (3%~10% noise)
> decompress:     no
> ignore binary:  no
> ignore files:   no
> index hidden:   no

  11539177677 67% Super-WPA.txt
 ------------ ---
  11539177677 67%

            1 files indexed in 1 directories
            1 new directories indexed
            1 new files indexed
            0 modified files indexed
            0 deleted files removed from indexes
            0 binary files indexed
            0 binary files ignored with --ignore-binary
            0 symbolic links skipped
            0 devices skipped
        65553 bytes indexing storage increase at 65553 bytes/file

Indexes are fresh and up to date
```

So now I try to grep something using index:
```
strace -f -s 111 ug --index -U  --stats "^echoeintro$" Super-WPA.txt 2>&1 | grep _Store
^C
```
and strace shows that `._UG#_Store` index files wasn't even opened. Why? Should I use ug in different way?

```
 ug -V
ugrep 6.5.0 x86_64-pc-linux-gnu +avx2; -P:pcre2jit; -z:zlib,bzip2,lzma,lz4,zstd,tar/pax/cpio/zip
License: BSD-3-Clause; ugrep user manual: <https://ugrep.com>
Written by Robert van Engelen and others: <https://github.com/Genivia/ugrep>
Ugrep utilizes the RE/flex regex library: <https://github.com/Genivia/RE-flex>
```

**Comment by marcin-github** [comment]
Also, UGStore contains only 0303 0707 bytes, is it correct?

**Comment by genivia-inc** [comment, marked as answer]
Index-based search is not used on a FILE target specified with ugrep. Only when searching recursively across a directory tree that contains index files.

**Comment by marcin-github** [comment]
Indeed, without specified file ugrep reads index files, thanks!
Btw, I still try to understand how index works. What kind  of data it keeps? And how it can speed up searching? I tried it on ~11GB text files and didn't notice any change of amount of data readed  from disk. What kind of regexp can be speedup by index?

**Comment by genivia-inc** [comment]
Index-based search works when the file system searched is cold or slow.

Indexing with lower accuracy (the default) is best, unless you have  regexes that MAY match a lot.

The more precise the regex, the faster the search is, because there are fewer false positives. False positives are also reduced with higher accuracy, but with `-9` you will get large index files that may not offer a speed benefit as much.

**Comment by marcin-github** [comment]
Hmm, file has 11G, index (created with -9 ) has 65k. And looks empty for me - this is something suspicious for me. 
Also I used regexp using "^fooo$" so only all line is searched, no wildcards but `ug --stats` shows
```
Searched 1 file in 1 directory in 99.5 seconds with 4 threads: 1 matching (100%)
Searched 982963903 lines: 3 matching (3.052e-07%)
Skipped 0 files of 0 (-nan%) not matching 1 indexes
```
ug found only three matching lines.
I'm missing something important in understanding how indexes in ug works but still can't get it.
// edit: i pasted wrong output of stats, now is correct

**Comment by genivia-inc** [comment]
The maximum index is 64K for any file, even when the file is very large.

This does not look suspicious. Large files are likely to have matches.
